File: core/MinimumHeightTrees/MinimumHeightTrees.py
from typing import List

def findMinHeightTrees(n: int, edges: List[List[int]]) -> List[int]:
    if n == 1:
        return [0]
    deg = [0] * n
    graph = [[] for _ in range(n)]
    for edge in edges:
        graph[edge[0]].append(edge[1])
        graph[edge[1]].append(edge[0])
        deg[edge[0]] += 1
        deg[edge[1]] += 1

    queue = [i for i in range(n) if deg[i] == 1]
    remain_num = n
    while remain_num > 2:
        remain_num -= len(queue)
        temp = queue
        queue = []
        for item in temp:
            for node in graph[item]:
                deg[node] -= 1
                if deg[node] == 1:
                    queue.append(node)
    return queue

File: core/MinimumHeightTrees/MinimumHeightTrees.java
package com.dbc;

import java.util.*;

public class MinimumHeightTrees {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (n == 1) return new ArrayList<Integer>(){{add(0);}};

        int[] edg = new int[n];
        Arrays.fill(edg, 0);
        List<Integer>[] graph = new List[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
            edg[edge[0]]++;
            edg[edge[1]]++;
        }

        List<Integer> queue = new ArrayList<>();
        for (int i = 0; i < edg.length; i++) {
            if (edg[i] == 1) queue.add(i);
        }
        while (n > 2) {
            n -= queue.size();
            List<Integer> temp = queue;
            queue = new ArrayList<>();
            for (int pos : temp) {
                for (int node : graph[pos]) {
                    edg[node]--;
                    if (edg[node] == 1) {
                        queue.add(node);
                    }
                }
            }
        }
        return queue;
    }
}
